### PR TITLE
Call `Strategy.process_dataloader` in `data_connector.py`

### LINKED ISSUE #12213: Refactor the call site of `strategy.process_dataloader()` from loop to data_connector
## Proposed refactor

This issue follows @ananthsub 's #11756 to move strategy-specific dataloader logic to the stategies. During the discussion, we noticed the existing `strategy.process_dataloader()`. 

`strategy.process_dataloader()` aims to provide additional strategy-specific data loading logic change. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/cf64f3443474a93d23b5afb0417e4a60298006e6/pytorch_lightning/strategies/strategy.py#L372-L378

Currently, `strategy.process_dataloader()` are called in `advance()` functions in  `fit_loop`, `eval_loop`, `prediction_loop`. 
For example, 
https://github.com/PyTorchLightning/pytorch-lightning/blob/6309a59c3cf93e0bfc352efb7cbf6c50b4544372/pytorch_lightning/loops/fit_loop.py#L275 

However, dataloader has been initialized in `_prepare_dataloader()` in data_connector.py, which is way before any loop starts. 
Therefore, we could refactor the call site of `strategy.process_dataloader()` by moving it to `_prepare_dataloader()` in data_connector.py. 

https://github.com/PyTorchLightning/pytorch-lightning/blob/a52a6ea0301abf93a288c27ff6297ec36ca7630d/pytorch_lightning/trainer/connectors/data_connector.py#L278-L285

By doing so, it has two benefits:
1. we could update dataloader before loop starts reading data. 
2. `advance()` was called multiple times within a loop, but strategy specific dataloader update should only need once. We could save unnecessary function calls. 

### Motivation

Currently, `strategy.process_dataloader()` are called in `advance()` functions in  `fit_loop`, `eval_loop`, `prediction_loop`. 
However, dataloader has been initialized in `_prepare_dataloader()` in data_connector.py, which is way before any loop starts. 
Therefore, we could refactor the call site of `strategy.process_dataloader()` by moving it to `_prepare_dataloader()` in data_connector.py. 

### Pitch

 We could refactor the call site of `strategy.process_dataloader()` by moving it to `_prepare_dataloader()` in data_connector.py. 

### Additional context

cc: @edward-io  @four4fish @ananthsub 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @ninginthecloud

### PR DESCRIPTION
**Pull request by ninginthecloud**
## What does this PR do?

Fixes #12213 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/data_connector.py
Comment (krshrimali) [inline pytorch_lightning/trainer/connectors/data_connector.py:316]: (I might be wrong here) - Do you think it's odd to call a public method from a assumably private method? I know private methods in Python are just those which start with `_` but still, asking out of curiosity.

**Review by krshrimali — COMMENTED** [review]
Thanks! I left a minor comment for you to take a look. Apart from that, if this PR looks good to everyone, do you think it's a good idea to update the function docstring with something like, "Calls the strategy's process_dataloader method"? (or anything that you feel is appropriate). This is the docstring I am referring to:

https://github.com/PyTorchLightning/pytorch-lightning/blob/7ca8c86c9d612363bc1895c8e1d01b5b4779e942/pytorch_lightning/trainer/connectors/data_connector.py#L277-L281

**Review by ananthsub — APPROVED** [review]
Please add an entry to the changelog too

File: pytorch_lightning/trainer/connectors/data_connector.py
Comment (ninginthecloud) [reply pytorch_lightning/trainer/connectors/data_connector.py:316]: Good callout~ Maybe code smell? But I think `DataConnector` has `trainer` property, which means we can get access to any public functions through `trainer`.  I think it's fine to call the public function from `Strategy` class, as long as it's not public function from `DataConnector`.  We need revisit this `_prepare_dataloader`, as I'm not sure why it's private func. It has been used for reset train/val dataloader frequently.

Comment (krshrimali) [reply pytorch_lightning/trainer/connectors/data_connector.py:317]: Thanks, @ninginthecloud for helping with the explanation behind this. Agreed that we can revisit the `_prepare_dataloader`, but let's not block this PR and get it landed if it looks good for everyone else as well. 🚀

File: CHANGELOG.md
Comment (krshrimali) [inline CHANGELOG.md:350]:
```suggestion
- Move `Strategy.process_dataloader` function call from `fit/evaluation/predict_loop.py` to `data_connector.py` ([#12251](https://github.com/PyTorchLightning/pytorch-lightning/pull/12251))
```

File: pytorch_lightning/trainer/connectors/data_connector.py
Comment (krshrimali) [inline pytorch_lightning/trainer/connectors/data_connector.py:281]: Thanks for adding this here! :))

**Review by krshrimali — APPROVED** [review]
LGTM, thanks @ninginthecloud!

File: CHANGELOG.md
Comment (ninginthecloud) [reply CHANGELOG.md:350]: Good catch! Thanks a lot for catching it!😃

**Comment by ninginthecloud** [comment]
Hi, code owners, @Borda, @SeanNaren, @carmocca , @tchaton , would you like to take a look of this PR and give an approval? Thanks a lot.

**Review by rohitgr7 — APPROVED** [review]

**Review by Borda — APPROVED** [review]

**Pull request merged**
